src/engine.ml

```diff
@@ -939,22 +939,48 @@ let server_handle_tls_data config auth_user_pass is_not_taken session keys tls d
       Ok (ip_config, config, channel_st, [ (`Control, out) ])
 
 let incoming_control_server auth_user_pass is_not_taken config rng session
-    channel _now _ts _key op data =
+    channel now _ts _key op data =
+  let open Result.Syntax in
   match (channel.channel_st, op) with
   | ( Expect_reset,
       ( Packet.Hard_reset_client_v2 | Packet.Hard_reset_client_v3
       | Packet.Soft_reset_v2 ) ) ->
-      (* TODO may need to do client certificate authentication here! *)
-      let _ca, server, key =
-        ( Config.get Ca config,
-          Config.get Tls_cert config,
-          Config.get Tls_key config )
+      let server, key =
+        (Config.get Tls_cert config, Config.get Tls_key config)
       in
       let ciphers = tls_ciphers config and version = tls_version config in
+      let* authenticator =
+        match
+          (Config.find Verify_client_cert config, Config.find Ca config)
+        with
+        | None, Some ca | Some `Required, Some ca ->
+            Ok
+              (Some
+                 (X509.Authenticator.chain_of_trust
+                    ~time:(fun () -> Some now)
+                    ~allowed_hashes:Mirage_crypto.Hash.hashes ca))
+        | Some `None, _ ->
+            Log.warn (fun m ->
+                m
+                  "server with 'verify-client-cert none', ensure a different \
+                   authentication mechanism is set up.");
+            Ok None
+        | Some `Optional, _ ->
+            Log.warn (fun m ->
+                m
+                  "server with 'verify-client-cert optional', ensure a \
+                   different authentication mechanism is set up.");
+            Ok (Some (fun ?ip:_ ~host:_ _certs -> Ok None))
```

Comment:
We should probably track authentication status somehow and try verify the cert. If it's signed by the CA then the client doesn't need to authenticate using username and password. I don't know if ocaml-tls allows for this kind of maybe-authentication... (I can investigate)

Reply:
Oh, I was not aware of the semantics of openvpn with `verify-tls-cert optional`. Is the following the case:
- the server requests a certificate
- if the ca or the peer fingerprint matches, we're done (i.e. no further authentication)
- if it does not succeed, we do username and password authentication (and depending on its success/failure, we proceed or fail establishing the connection)

The only path I can think of right now is to have a `mutable authenticated : bool` in the type t, and set this from the authenticator callback. What do you think?

OCaml-TLS API is: you can provide an `authenticator`, which returns Ok or Error. On Ok, the TLS handshake is done further, on Error the handshake errors (and thus you don't get a TLS session).

Reply:
Yes, mostly.

* The server requests a certificate
* if the CA or peer-fingerprint matches the client is authenticated
* the client can send user-pass regardless of certificate authentication status and is (double) authenticated if the user-pass is valid

All this gets more complicated with all the callback scripts you can hook in into the OpenVPN authentication process. I don't know what you can do with the callback scripts and I'm afraid to look :D

> The only path I can think of right now is to have a mutable authenticated : bool in the type t, and set this from the authenticator callback. What do you think?

Hmm in that case we should be careful because we create new `t`s all the time.

Looking at the code the `authenticator` returns either `Ok None` or `Ok (Some leaf_cert_or_is_it_cert_chain)`. If so we can maybe check after the handshake if the tls object has a peer certificate and in that case consider the peer authenticated.

Reply:
Yes, it's a big footgun. We could as well in `Miragevpn.server` fail with an error on `verify-tls-cert optional` and `auth_user_pass = None`.

I think the `verify-tls-cert optional` could be useful if you have some (e.g. headless) clients where a certificate is more suitable for authentication mixed with clients with a human user behind that can input username and password (but can't remember their certificate and key).

So it's not clear to me if we want to support `verify-tls-cert optional` or not.